### PR TITLE
Differential expressions

### PR DESCRIPTION
**Pull request by zigasinko**
Opened an issue on Vega for slow selection rect drawing if xaxis label orientation is `bottom`. Will change it once they fix the bug: https://github.com/vega/vega/issues/2927

### CODE REVIEW

File: src/components/app/app.tsx [hunk 1]
```diff
@@ -14,34 +14,36 @@ import { SnackbarProvider } from 'notistack';
...
-                                    <Route path="/bcm" component={GeneExpressGrid} />
-                                    <Route component={PageNotFound} />
-                                </Switch>
-                            </BrowserRouter>
-                        </SnackbarProvider>
-                    </StyledComponentsThemeProvider>
-                </MuiThemeProvider>
-            </ReduxProvider>
-        </StylesProvider>
+        <RendererContext.Provider value="canvas">
```
Comment (dblenkus) [inline src/components/app/app.tsx:26]: Why don't you save this into the Redux?

File: src/components/genexpress/common/chart/chart.tsx [hunk 2]
```diff
@@ -55,23 +66,23 @@ const Chart = ({
     /**
      * Retrieve width of chart element (chart container div).
      */
-    const getChartWidth = useCallback((): number => {
+    const getChartElementWidth = useCallback((): number => {
         if (!chartElement.current) {
             return 0;
         }
 
-        return chartElement.current.clientWidth - chartPadding.left - chartPadding.right;
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx:63]: These changes in the layout should go into a separate commit, not under differential expressions.

File: src/components/genexpress/common/chart/chart.tsx [hunk 3]
```diff
@@ -90,16 +101,16 @@ const Chart = ({
...
 
         runChart();
-    }, [getChartHeight, getChartWidth, runChart]);
+    }, [getChartElementHeight, getChartElementWidth, runChart]);
 
     /**
      * Resize on parent element width / height change.
      */
-    useEffect(() => {
+    useUpdateEffect(() => {
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx:113]: You could also put this into a separate commit.

File: src/components/genexpress/common/chart/chart.tsx [hunk 4]
```diff
@@ -146,7 +155,7 @@ const Chart = ({
             // registered on external DOM elements.
             chartView.current?.finalize();
         };
-    }, [getChartHeight, getChartWidth, runChart, vegaSpecification]);
+    }, [getChartElementHeight, getChartElementWidth, renderer, runChart, vegaSpecification]);
```
Comment (dblenkus) [inline src/components/genexpress/common/chart/chart.tsx:158]: Why is `renderer` added only now? Doesn't linters detect such errors?

File: src/components/genexpress/geneExpressGrid.tsx [hunk 5]
```diff
@@ -123,7 +166,7 @@ const GeneExpressGrid = ({
                 draggableHandle=".dragHandle"
                 layouts={defaultLayout}
                 verticalCompact
-                breakpoints={{ lg: 1200, md: 996, sm: 768 }}
+                breakpoints={{ lg: breakpoints.big, md: breakpoints.mid, sm: breakpoints.small }}
```
Comment (dblenkus) [inline src/components/genexpress/geneExpressGrid.tsx:169]: There is a lot of changes in this file that could go into separate commits.

File: src/components/genexpress/modules/differentialExpressions/differentialExpressions.tsx [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+const mapStateToProps = (
+    state: RootState,
+): {
+    differentialExpressions: DifferentialExpression[];
+    selectedDifferentialExpression: DifferentialExpression;
+    highlightedGenesIds: string[];
+    selectedGenesIds: string[];
+} => {
+    return {
+        // All available differential expressions
```
Comment (dblenkus) [inline src/components/genexpress/modules/differentialExpressions/differentialExpressions.tsx]: Add a dot at the end.

File: src/redux/epics/timeSeriesEpics.ts [hunk 7]
```diff
@@ -105,22 +120,15 @@ export const fetchTimeSeriesSamplesExpressionsEpic: Epic<Action, Action, RootSta
         mergeMap(([, state]) => {
             const timeSeriesSamplesExpressions = {} as SamplesExpressionsById;
 
-            const selectedTimeSeriesId = getSelectedTimeSeries(state.timeSeries);
-
-            const timeSeriesSamplesIds = getTimeSeriesSamplesIds(
-                selectedTimeSeriesId.id,
-                state.timeSeries,
-            );
+            const timeSeriesSamplesIds = getSelectedTimeSeriesSamplesIds(state.timeSeries);
```
Comment (dblenkus) [inline src/redux/epics/timeSeriesEpics.ts:123]: Is this change in the right commit?

File: src/redux/models/internal.ts [hunk 8]
```diff
@@ -35,6 +41,41 @@ export type SamplesExpressionsById = {
...
+    status: DataStatus;
+    started: string;
+    finished: string;
+    process_progress: number;
+    logfc_threshold: number;
+    prob_field: string;
+    prob_threshold: number;
+    up_regulated: number;
+    down_regulated: number;
+    output: DataDifferentialExpression['output'];
```
Comment (dblenkus) [inline src/redux/models/internal.ts]: Why do you use `DataDifferentialExpression` only for a single attribute instead of the whole type?

File: src/components/app/app.tsx (hunk 1 above)
Comment (zigasinko) [reply src/components/app/app.tsx:26]: I could but I think it doesn't belong there because visualizations are always displayed on canvas. SVG value is only used in tests. So it's more of a system setting than a datum for redux store.

File: src/components/genexpress/common/chart/chart.tsx (hunk 4 above)
Comment (zigasinko) [reply src/components/genexpress/common/chart/chart.tsx:158]: Because it was a `const` before and it never changed. Yes, linters `react-hooks/exhaustive-deps` rule throws an error about this.

File: src/components/genexpress/modules/differentialExpressions/differentialExpressions.tsx (hunk 6 above)
Comment (zigasinko) [reply src/components/genexpress/modules/differentialExpressions/differentialExpressions.tsx]: Added a dot.

File: src/redux/models/internal.ts (hunk 8 above)
Comment (zigasinko) [reply src/redux/models/internal.ts]: Migrated this type from genialis-apps. Now I've fixed it to use the whole type and added only missing properties.

File: src/components/genexpress/common/chart/chart.tsx (hunk 3 above)
Comment (zigasinko) [reply src/components/genexpress/common/chart/chart.tsx:113]: Chart changes moved to a separate commit `Add updatableSignalDefinitions and renderer context to chart`.

File: src/components/genexpress/common/chart/chart.tsx (hunk 2 above)
Comment (zigasinko) [reply src/components/genexpress/common/chart/chart.tsx:63]: Chart changes moved to a separate commit `Add updatableSignalDefinitions and renderer context to chart`.

File: src/components/genexpress/geneExpressGrid.tsx (hunk 5 above)
Comment (zigasinko) [reply src/components/genexpress/geneExpressGrid.tsx:169]: Changes moved to a separate commit `Fix geneExpressGrid breakpoints and expressions module id`.

File: src/redux/epics/timeSeriesEpics.ts (hunk 7 above)
Comment (zigasinko) [reply src/redux/epics/timeSeriesEpics.ts:123]: Changes moved to a separate commit `Refactor time series epics handling samples ids`.

**Review by dblenkus — APPROVED** [review]
